File: tests/test_timer.py
"""Tests for the timer module."""

import time
import pytest
from termino.timer import Stopwatch, Countdown, TimerState


class TestStopwatch:
    """Tests for the Stopwatch class."""

    def test_initial_state(self):
        """A new stopwatch should be IDLE with zero elapsed time."""
        sw = Stopwatch()
        assert sw.state == TimerState.IDLE
        assert sw.elapsed == 0.0
        assert sw.laps == []

    def test_start_transitions_to_running(self):
        """Starting the stopwatch should set state to RUNNING."""
        sw = Stopwatch()
        sw.start()
        assert sw.state == TimerState.RUNNING

    def test_elapsed_increases_while_running(self):
        """Elapsed time should increase while the stopwatch is running."""
        sw = Stopwatch()
        sw.start()
        time.sleep(0.05)
        elapsed = sw.elapsed
        assert elapsed > 0.0
        # Allow some tolerance
        assert 0.01 < elapsed < 0.5

    def test_pause_stops_elapsed(self):
        """Pausing should stop the elapsed time from increasing."""
        sw = Stopwatch()
        sw.start()
        time.sleep(0.05)
        sw.pause()
        paused_elapsed = sw.elapsed
        time.sleep(0.05)
        # Should not increase while paused
        assert sw.elapsed == paused_elapsed
        assert sw.state == TimerState.PAUSED

    def test_resume_continues_elapsed(self):
        """Resuming should continue the elapsed time."""
        sw = Stopwatch()
        sw.start()
        time.sleep(0.05)
        sw.pause()
        paused_elapsed = sw.elapsed
        sw.resume()
        assert sw.state == TimerState.RUNNING
        time.sleep(0.05)
        assert sw.elapsed > paused_elapsed

    def test_stop_returns_elapsed(self):
        """Stopping should return the elapsed time."""
        sw = Stopwatch()
        sw.start()
        time.sleep(0.05)
        final = sw.stop()
        assert sw.state == TimerState.STOPPED
        assert final > 0.0
        # Elapsed should not change after stop
        assert sw.elapsed == final

    def test_lap_recording(self):
        """Laps should be recorded correctly."""
        sw = Stopwatch()
        sw.start()
        time.sleep(0.05)
        lap1 = sw.lap()
        time.sleep(0.05)
        lap2 = sw.lap()
        sw.stop()

        assert len(sw.laps) == 2
        assert lap1.lap == 1
        assert lap2.lap == 2
        assert lap2.total > lap1.total
        assert lap2.split > 0.0

    def test_lap_while_not_running_raises(self):
        """Recording a lap while not running should raise RuntimeError."""
        sw = Stopwatch()
        with pytest.raises(RuntimeError):
            sw.lap()

    def test_pause_while_not_running_raises(self):
        """Pausing while not running should raise RuntimeError."""
        sw = Stopwatch()
        with pytest.raises(RuntimeError):
            sw.pause()

    def test_resume_while_not_paused_raises(self):
        """Resuming while not paused should raise RuntimeError."""
        sw = Stopwatch()
        with pytest.raises(RuntimeError):
            sw.resume()

    def test_start_while_running_raises(self):
        """Starting while already running should raise RuntimeError."""
        sw = Stopwatch()
        sw.start()
        with pytest.raises(RuntimeError):
            sw.start()

    def test_start_after_stop(self):
        """Starting after stopping should reset the stopwatch."""
        sw = Stopwatch()
        sw.start()
        time.sleep(0.05)
        sw.stop()
        sw.start()
        assert sw.state == TimerState.RUNNING
        time.sleep(0.05)
        assert sw.elapsed > 0.0


class TestCountdown:
    """Tests for the Countdown class."""

    def test_initial_state(self):
        """A new countdown should be IDLE with full remaining time."""
        cd = Countdown(duration=10.0)
        assert cd.state == TimerState.IDLE
        assert cd.remaining == 10.0
        assert cd.elapsed == 0.0
        assert not cd.is_finished

    def test_start_transitions_to_running(self):
        """Starting should set state to RUNNING."""
        cd = Countdown(duration=10.0)
        cd.start()
        assert cd.state == TimerState.RUNNING

    def test_remaining_decreases_while_running(self):
        """Remaining time should decrease while running."""
        cd = Countdown(duration=1.0)
        cd.start()
        time.sleep(0.2)
        assert cd.remaining < 1.0
        assert cd.elapsed > 0.0

    def test_is_finished_when_time_expires(self):
        """is_finished should be True when time is up."""
        cd = Countdown(duration=0.1)
        cd.start()
        time.sleep(0.2)
        assert cd.is_finished
        assert cd.remaining == 0.0

    def test_pause_stops_countdown(self):
        """Pausing should stop the countdown from decreasing."""
        cd = Countdown(duration=5.0)
        cd.start()
        time.sleep(0.05)
        cd.pause()
        remaining_paused = cd.remaining
        time.sleep(0.05)
        assert cd.remaining == remaining_paused

    def test_resume_continues_countdown(self):
        """Resuming should continue the countdown."""
        cd = Countdown(duration=5.0)
        cd.start()
        time.sleep(0.05)
        cd.pause()
        remaining_after_pause = cd.remaining
        cd.resume()
        time.sleep(0.05)
        assert cd.remaining < remaining_after_pause

    def test_stop_returns_elapsed(self):
        """Stopping should return the elapsed time."""
        cd = Countdown(duration=5.0)
        cd.start()
        time.sleep(0.05)
        final = cd.stop()
        assert cd.state == TimerState.STOPPED
        assert final > 0.0

    def test_start_while_running_raises(self):
        """Starting while already running should raise RuntimeError."""
        cd = Countdown(duration=5.0)
        cd.start()
        with pytest.raises(RuntimeError):
            cd.start()

    def test_pause_while_not_running_raises(self):
        """Pausing while not running should raise RuntimeError."""
        cd = Countdown(duration=5.0)
        with pytest.raises(RuntimeError):
            cd.pause()

    def test_resume_while_not_paused_raises(self):
        """Resuming while not paused should raise RuntimeError."""
        cd = Countdown(duration=5.0)
        with pytest.raises(RuntimeError):
            cd.resume()

File: tests/test_timer.rs
use std::time::Duration;

use termino::timer::{Countdown, Stopwatch, TimerState};

#[test]
fn test_stopwatch_initial_state() {
    let sw = Stopwatch::new();
    assert_eq!(sw.state(), TimerState::Idle);
    assert_eq!(sw.elapsed(), 0.0);
    assert!(sw.laps().is_empty());
}

#[test]
fn test_stopwatch_start_transitions_to_running() {
    let mut sw = Stopwatch::new();
    sw.start().unwrap();
    assert_eq!(sw.state(), TimerState::Running);
}

#[test]
fn test_stopwatch_elapsed_increases_while_running() {
    let mut sw = Stopwatch::new();
    sw.start().unwrap();
    std::thread::sleep(Duration::from_millis(50));
    let elapsed = sw.elapsed();
    assert!(elapsed > 0.0);
    assert!(elapsed < 0.5);
}

#[test]
fn test_stopwatch_pause_stops_elapsed() {
    let mut sw = Stopwatch::new();
    sw.start().unwrap();
    std::thread::sleep(Duration::from_millis(50));
    sw.pause().unwrap();
    let paused_elapsed = sw.elapsed();
    std::thread::sleep(Duration::from_millis(50));
    assert_eq!(sw.elapsed(), paused_elapsed);
    assert_eq!(sw.state(), TimerState::Paused);
}

#[test]
fn test_stopwatch_resume_continues_elapsed() {
    let mut sw = Stopwatch::new();
    sw.start().unwrap();
    std::thread::sleep(Duration::from_millis(50));
    sw.pause().unwrap();
    let paused_elapsed = sw.elapsed();
    sw.resume().unwrap();
    assert_eq!(sw.state(), TimerState::Running);
    std::thread::sleep(Duration::from_millis(50));
    assert!(sw.elapsed() > paused_elapsed);
}

#[test]
fn test_stopwatch_stop_returns_elapsed() {
    let mut sw = Stopwatch::new();
    sw.start().unwrap();
    std::thread::sleep(Duration::from_millis(50));
    let final_time = sw.stop();
    assert_eq!(sw.state(), TimerState::Stopped);
    assert!(final_time > 0.0);
    assert_eq!(sw.elapsed(), final_time);
}

#[test]
fn test_stopwatch_lap_recording() {
    let mut sw = Stopwatch::new();
    sw.start().unwrap();
    std::thread::sleep(Duration::from_millis(30));
    let lap1 = sw.lap().unwrap();
    std::thread::sleep(Duration::from_millis(30));
    let lap2 = sw.lap().unwrap();
    sw.stop();

    assert_eq!(sw.laps().len(), 2);
    assert_eq!(lap1.lap, 1);
    assert_eq!(lap2.lap, 2);
    assert!(lap2.total > lap1.total);
    assert!(lap2.split > 0.0);
}

#[test]
fn test_stopwatch_lap_while_not_running_raises() {
    let mut sw = Stopwatch::new();
    assert!(sw.lap().is_err());
}

#[test]
fn test_stopwatch_pause_while_not_running_raises() {
    let mut sw = Stopwatch::new();
    assert!(sw.pause().is_err());
}

#[test]
fn test_stopwatch_resume_while_not_paused_raises() {
    let mut sw = Stopwatch::new();
    assert!(sw.resume().is_err());
}

#[test]
fn test_stopwatch_start_while_running_raises() {
    let mut sw = Stopwatch::new();
    sw.start().unwrap();
    assert!(sw.start().is_err());
}

#[test]
fn test_stopwatch_start_after_stop() {
    let mut sw = Stopwatch::new();
    sw.start().unwrap();
    std::thread::sleep(Duration::from_millis(30));
    sw.stop();
    sw.start().unwrap();
    assert_eq!(sw.state(), TimerState::Running);
    std::thread::sleep(Duration::from_millis(30));
    assert!(sw.elapsed() > 0.0);
}

#[test]
fn test_countdown_initial_state() {
    let cd = Countdown::new(10.0);
    assert_eq!(cd.state(), TimerState::Idle);
    assert_eq!(cd.remaining(), 10.0);
    assert_eq!(cd.elapsed(), 0.0);
    assert!(!cd.is_finished());
}

#[test]
fn test_countdown_start_transitions_to_running() {
    let mut cd = Countdown::new(10.0);
    cd.start().unwrap();
    assert_eq!(cd.state(), TimerState::Running);
}

#[test]
fn test_countdown_remaining_decreases_while_running() {
    let mut cd = Countdown::new(1.0);
    cd.start().unwrap();
    std::thread::sleep(Duration::from_millis(200));
    assert!(cd.remaining() < 1.0);
}

#[test]
fn test_countdown_is_finished_when_time_expires() {
    let mut cd = Countdown::new(0.1);
    cd.start().unwrap();
    std::thread::sleep(Duration::from_millis(200));
    assert!(cd.is_finished());
    assert_eq!(cd.remaining(), 0.0);
}

#[test]
fn test_countdown_pause_stops_countdown() {
    let mut cd = Countdown::new(5.0);
    cd.start().unwrap();
    std::thread::sleep(Duration::from_millis(50));
    cd.pause().unwrap();
    let remaining_paused = cd.remaining();
    std::thread::sleep(Duration::from_millis(50));
    assert_eq!(cd.remaining(), remaining_paused);
}

#[test]
fn test_countdown_resume_continues_countdown() {
    let mut cd = Countdown::new(5.0);
    cd.start().unwrap();
    std::thread::sleep(Duration::from_millis(50));
    cd.pause().unwrap();
    let remaining_after_pause = cd.remaining();
    cd.resume().unwrap();
    std::thread::sleep(Duration::from_millis(50));
    assert!(cd.remaining() < remaining_after_pause);
}

#[test]
fn test_countdown_stop_returns_elapsed() {
    let mut cd = Countdown::new(5.0);
    cd.start().unwrap();
    std::thread::sleep(Duration::from_millis(50));
    let final_time = cd.stop();
    assert_eq!(cd.state(), TimerState::Stopped);
    assert!(final_time > 0.0);
}

#[test]
fn test_countdown_start_while_running_raises() {
    let mut cd = Countdown::new(5.0);
    cd.start().unwrap();
    assert!(cd.start().is_err());
}

#[test]
fn test_countdown_pause_while_not_running_raises() {
    let mut cd = Countdown::new(5.0);
    assert!(cd.pause().is_err());
}

#[test]
fn test_countdown_resume_while_not_paused_raises() {
    let mut cd = Countdown::new(5.0);
    assert!(cd.resume().is_err());
}